main.cpp

```diff
@@ -89,422 +89,222 @@ If you need to view an example, see: https://bitbucket.org/MatkatMusic/pfmcpptas
-}
-
-IntType& IntType::pow(const FloatType& floatRef)
-{
-    *intTypeHeap = (powInternal(static_cast<int>(floatRef)));
-    return *this;
-}
-
-//FloatType Implementation //////////////////////////////////////
-
-
-FloatType::FloatType(float floatValue)
-{
-    floatTypeHeap = new float(floatValue);
-}
-
-FloatType::~FloatType()
-{
-    delete floatTypeHeap;
-}
-
-FloatType& FloatType::operator+=( const float other) 
-{
-    *floatTypeHeap += other;
-    return *this;
-}
-
-FloatType& FloatType::operator-=(const float other)
-{
-    *floatTypeHeap -= other;
-    return *this;
-}
 
-FloatType& FloatType::operator/=( const float other) 
-{
-    *floatTypeHeap /= other;
-    return *this;
-}
-
-FloatType& FloatType::operator*=(const float other) 
-{
-    *floatTypeHeap *= other;
-    return *this;
-}
-
-FloatType& FloatType::apply(std::function<FloatType&(float&)> floatFunc)
-{
-    if(floatFunc)
+    Numeric& operator/=(const MyType other)
     {
-        return floatFunc(*floatTypeHeap);
+        // if your template type is an int
+        // if your parameter's type is also an int
+        //                 if your parameter is 0
+        //                         don't do the division
+        //         else if it's less than epsilon
+        //                 dont do the divison
+        // else if it's less than epsilon
+        //         warn about doing the division
+        if(std::is_same<int, MyType>::value)
+        {
+            if(std::is_same<decltype(other), int>::value)
+            {
+                if(other == 0.f)
+                {
+                    return *this;
+                }
+                else if(std::numeric_limits<MyType>::is_integer)
+                {
+                    return *this;
+                }
+                else
+                {
+                    std::cout<<"Warning division by zero";
+                }
+
+            } 
+        } 
+        return *this;
     }
-    return *this;
-}
 
-FloatType& FloatType::apply(void(*floatFunc)(float&))
-{
-    if(floatFunc)
+    Numeric& operator*=(const MyType other)
     {
-        floatFunc(*floatTypeHeap);
+        *numericHeap *= other;
+        return *this;
+    }
+    template<typename X>
+    Numeric& apply(X numericFunc)
+    {
+        numericFunc(numericHeap);
+        return *this;
     }
-    return *this;
-}
-
-FloatType& FloatType::pow(float power)
-{
-    powInternal(power);
-    return *this;
-}
-
-FloatType& FloatType::pow(const IntType& intRef)
-{ 
-    powInternal(intRef);
-    return *this; 
-}
-
-FloatType& FloatType::pow(const FloatType& floatRef)
-{
-    powInternal(floatRef);
-    return *this; 
-}
 
-FloatType& FloatType::pow(const DoubleType& doubleRef)
-{
-    *floatTypeHeap = powInternal(static_cast<float>(doubleRef)); 
-    return *this; 
-}
+    
+    Numeric& apply(void(*numericFunc)(std::unique_ptr<MyType>&))
+    {
+        numericFunc(numericHeap);
+        return *this;
+    }
```

Comment:
You're not checking if `numericFunc` is a `nullptr` or not
- [ ] revert to previous function ptr version of `apply` and revise from there.